hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.query.sqm.tree.expression;
+
+import org.hibernate.query.sqm.SemanticQueryWalker;
+import org.hibernate.query.sqm.SqmExpressible;
+import org.hibernate.query.sqm.tree.SqmCopyContext;
+
+public class AsWrapperSqmExpression<T> extends AbstractSqmExpression<T> {
```

Comment:
I'd personally prefer naming this `TreatExpression` and use the same `treat(.. as type)` syntax that we use for `from` nodes. Wdyt @sebersole ?

Reply:
I think I'd prefer to avoid the confusion with TREAT

Reply:
So do we want to support a HQL syntax for this as well? Or would you say this is a JPA Criteria only concept?

Reply:
I'm not seeing any benefit to this in HQL.  Is there something I'm missing?

Reply:
Here comes a use case that just came to mind. Projecting something that is of type `java.sql.Timestamp` in a constructor to `Instant` via e.g.:

```
select new MyDto(e.id, treat(e.created as Instant))
from MyEntity e
```

Reply:
Yeah, right, you might want to cast the "HQL" (i.e. Java) type to make HQL's type system happy, even though at the database level there is no work to do because the two Java types have already been erased down to the same SQL type.

It actually does make sense.

Reply:
Here a few suggestions for function names:

* `morph(timestampProperty as Instant)`
* `transform(timestampProperty as Instant)`
* `transmute(timestampProperty as Instant)`
* `cast_optional(timestampProperty as Instant)`

Here a few names I dismissed because SQL functions with the same name exist:

* `convert` 
* `translate`